docs/rfcs/041-sharded-ingest.md

```diff
@@ -0,0 +1,267 @@
+depth.
+
+### Perhaps: Serve reads from the safekeeper/eliminate pageserver InMemoryLayer
+
+Now that we are doing some pre-processing of the WAL on the safekeeper and storing
+a buffer of some recent reads in Cursor, it would be possible to serve some reads
+from the safekeeper directly.  It's unclear whether this is a desirable change: it would
+only be a net benefit if it enabled us to avoid doing some work on the pageserver.
+
+Ultimately, if the safekeeper served recent reads, we could build entire L0 layers on the safekeeper
+and upload them to S3, saving the pageserver the disk bandwidth of writing ephemeral layers
+and then flushing them to L0s, since ephemeral layers are approximately just extents of WAL, post-processed
+into a readable form.
+
+## Alternatives considered
+
+### Give safekeepers enough state to fully decode WAL
+
+In this RFC, we only do the first phase of ingest on the safekeeper, because this is
+the phase that is stateless.  Subsequent changes then happen on the pageserver, with
+access to the `Timeline` state.
+
+We could do more work on the safekeeper if we transmitted metadata state to the safekeeper
+when subscribing to the WAL: for example, by telling the safekeeper all the relation sizes,
+so that it could then generate all the metadata writes for relation sizes.
+
+We avoid doing this for several reasons:
+1. Complexity: it's a more invasive protocol change
+2. Decoupling: having the safekeeper understand the `ProcessedWalIngest` already somewhat
+   infects it with knowledge of the pageserver, but this is mainly an abstract structure
+   that describes postgres writes.  However, if we taught the safekeeper about the exact
+   way that pageserver deals with metadata keys, this would be a much tighter coupling.
+3. Load: once the WAL has been processed to the point that it can be split between shards,
+   it is preferable to share out work on the remaining shards rather than adding extra CPU
+   load to the safekeeper.
+
+### Do pre-processing on the compute instead of the safekeeper
+
+Since our first stage of ingest is stateless, it could be done at any stage in the pipeline,
+all the way up to the compute.
+
+We choose not to do this, because it is useful for the safekeeper to store the raw WAL rather
+than just the preprocessed WAL:
+- The safekeeper still needs to be able to serve raw WAL back to postgres for e.g. physical replication
+- It simplifies our paxos implementation to have the offset in the write log be literally
+  the same as the LSN
+
+### Do wal pre-processing on shard 0 or a separate service, send it to other shards from there
+
+If we wanted to keep the safekeepers as entirely pure stores of raw WAL bytes, then
+we could do the initial decode and shard-splitting in some other location:
+- Shard 0 could subscribe to the full WAL and then send writes to other shards
+- A new intermediate service between the safekeeper and pageserver could do the splitting.
+
+So why not?
```

Comment:
It also adds an extra network hop from shard 0 to the final shard. I think that's reason enough not to do this.

Reply:
https://github.com/neondatabase/neon/pull/8754/commits/818965ff7ff80ee1d5a07b5b038a986708520a7e